File: JavaScript/6-LinkedList/Problems/3-LinkedListCycle/LinkedListCyclePro.js
// Definition of a ListNode (you can assume this is given)
class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

// Helper function to create a linked list from an array
function createLinkedList(arr, pos = -1) {
  if (!arr || arr.length === 0) {
    return { head: null, tail: null };
  }

  const head = new ListNode(arr[0]);
  let current = head;
  let tail = head;
  let cycleStart = null;

  if (pos === 0) {
    cycleStart = head;
  }

  for (let i = 1; i < arr.length; i++) {
    const newNode = new ListNode(arr[i]);
    current.next = newNode;
    current = newNode;
    tail = newNode;
    if (i === pos) {
      cycleStart = newNode;
    }
  }

  if (cycleStart) {
    tail.next = cycleStart;
  }

  return { head, tail };
}

// Helper function to visualize linked list with cycle
function visualizeLinkedList(head, highlightCycle = false) {
  let current = head;
  let nodes = [];
  let visited = new Set();
  let cycleStartNode = null;

  while (current) {
    if (visited.has(current)) {
      cycleStartNode = current;
      break; // Break, but we still want to process the last node
    }
    nodes.push(current.val);
    visited.add(current);
    current = current.next;
  }

  let displayString = nodes.join(" -> ");
  if (cycleStartNode && highlightCycle) {
        let cycleStartIndex = nodes.indexOf(cycleStartNode.val);
        let cycleString = nodes.slice(cycleStartIndex).join(" -> ");
        displayString = displayString.substring(0, displayString.indexOf(cycleString[0])) + "[" + cycleString + "]";
  } else if (cycleStartNode) {
        let cycleStartIndex = nodes.indexOf(cycleStartNode.val);
        let cycleString = nodes.slice(cycleStartIndex).join(" -> ");
        displayString = displayString.substring(0, displayString.indexOf(cycleString[0])) + "[" + cycleString + "...(cycle)]";
  }


  return displayString;
}

// 1. Floyd's Cycle Detection (Fast and Slow Pointers) - Most common and efficient
function hasCycleFloyd(head) {
  if (!head || !head.next) {
    return false;
  }
  let slow = head;
  let fast = head.next;
  while (slow !== fast) {
    if (!fast || !fast.next) {
      return false;
    }
    slow = slow.next;
    fast = fast.next.next;
  }
  return true;
}

// 2. Hashing (Using a Set) - Simple, but uses extra space
function hasCycleHashing(head) {
  const visited = new Set();
  let current = head;
  while (current) {
    if (visited.has(current)) {
      return true;
    }
    visited.add(current);
    current = current.next;
  }
  return false;
}

// 3. Modifying the List (Not recommended for real-world, alters original list)
function hasCycleModify(head) {
  let current = head;
  while (current) {
    if (current.visited) {
      return true;
    }
    current.visited = true; // Add a visited property
    current = current.next;
  }
  return false;
}

// 4. Using an Array to Store Nodes - Another approach using extra space
function hasCycleArray(head) {
  const nodes = [];
  let current = head;
  while (current) {
    if (nodes.includes(current)) {
      return true;
    }
    nodes.push(current);
    current = current.next;
  }
  return false;
}

// 5. Recursive Approach (Not ideal for very long lists due to call stack)
function hasCycleRecursive(node, visited = new Set()) {
  if (!node) {
    return false;
  }
  if (visited.has(node)) {
    return true;
  }
  visited.add(node);
  return hasCycleRecursive(node.next, visited);
}

// Test cases
function runTestCases() {
  let list1 = createLinkedList([1, 2, 3, 4, 5]);
  let list2 = createLinkedList([1, 2, 3, 4, 5], 2); // Cycle starts at 3
  let list3 = createLinkedList([1, 2]);
  let list4 = createLinkedList([1, 2], 0); // Cycle starts at 1
  let list5 = createLinkedList([1]);
  let list6 = createLinkedList([]);
  let list7 = createLinkedList([1,2,3,4,5,6,7,8,9,10], 5); // Cycle at 6th element

  console.log("Test Case 1: No cycle");
  console.log("Input: ", visualizeLinkedList(list1.head));
  console.log("Floyd's:", hasCycleFloyd(list1.head));
  console.log("Hashing:", hasCycleHashing(list1.head));
  console.log("Modifying:", hasCycleModify(list1.head));
  console.log("Array:", hasCycleArray(list1.head));
  console.log("Recursive:", hasCycleRecursive(list1.head));
  console.log("------------------");

  // Reset visited property for modify approach
  list1 = createLinkedList([1, 2, 3, 4, 5]);

  console.log("Test Case 2: Cycle in the middle");
  console.log("Input: ", visualizeLinkedList(list2.head, true));
  console.log("Floyd's:", hasCycleFloyd(list2.head));
  console.log("Hashing:", hasCycleHashing(list2.head));
  console.log("Modifying:", hasCycleModify(list2.head));
  console.log("Array:", hasCycleArray(list2.head));
  console.log("Recursive:", hasCycleRecursive(list2.head));
  console.log("------------------");

    // Reset visited property for modify approach
    list2 = createLinkedList([1, 2, 3, 4, 5], 2);

  console.log("Test Case 3: No cycle, short list");
  console.log("Input: ", visualizeLinkedList(list3.head));
  console.log("Floyd's:", hasCycleFloyd(list3.head));
  console.log("Hashing:", hasCycleHashing(list3.head));
  console.log("Modifying:", hasCycleModify(list3.head));
  console.log("Array:", hasCycleArray(list3.head));
  console.log("Recursive:", hasCycleRecursive(list3.head));
  console.log("------------------");

    // Reset visited property for modify approach
    list3 = createLinkedList([1, 2]);

  console.log("Test Case 4: Cycle at the beginning");
  console.log("Input: ", visualizeLinkedList(list4.head, true));
  console.log("Floyd's:", hasCycleFloyd(list4.head));
  console.log("Hashing:", hasCycleHashing(list4.head));
  console.log("Modifying:", hasCycleModify(list4.head));
  console.log("Array:", hasCycleArray(list4.head));
  console.log("Recursive:", hasCycleRecursive(list4.head));
  console.log("------------------");

    list4 = createLinkedList([1, 2], 0);

  console.log("Test Case 5: Single node, no cycle");
  console.log("Input: ", visualizeLinkedList(list5.head));
  console.log("Floyd's:", hasCycleFloyd(list5.head));
  console.log("Hashing:", hasCycleHashing(list5.head));
  console.log("Modifying:", hasCycleModify(list5.head));
  console.log("Array:", hasCycleArray(list5.head));
  console.log("Recursive:", hasCycleRecursive(list5.head));
  console.log("------------------");

  list5 = createLinkedList([1]);

  console.log("Test Case 6: Empty list");
  console.log("Input: ", visualizeLinkedList(list6.head));
  console.log("Floyd's:", hasCycleFloyd(list6.head));
  console.log("Hashing:", hasCycleHashing(list6.head));
  console.log("Modifying:", hasCycleModify(list6.head));
  console.log("Array:", hasCycleArray(list6.head));
  console.log("Recursive:", hasCycleRecursive(list6.head));
  console.log("------------------");

  list6 = createLinkedList([]);

  console.log("Test Case 7: Larger list with cycle");
  console.log("Input: ", visualizeLinkedList(list7.head, true));
  console.log("Floyd's:", hasCycleFloyd(list7.head));
  console.log("Hashing:", hasCycleHashing(list7.head));
  console.log("Modifying:", hasCycleModify(list7.head));
  console.log("Array:", hasCycleArray(list7.head));
  console.log("Recursive:", hasCycleRecursive(list7.head));
  console.log("------------------");
}

runTestCases();

File: C++/6-LinkedList/Problems/3-LinkedListCycle/LinkedListCyclePro.cpp
#include <iostream>
#include <unordered_set>
#include <vector>
#include <algorithm>
#include <climits> // Include climits for INT_MIN

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

// 1. Floyd's Cycle Detection Algorithm (Fast and Slow Pointers)
//   - Most common and efficient approach.
//   - Real-world application: Detecting loops in network routing tables.  If a router's table has a loop, packets can get stuck.
bool hasCycle_Floyd(ListNode *head) {
    if (!head || !head->next) {
        return false;
    }
    ListNode *slow = head;
    ListNode *fast = head->next; // Start fast pointer one step ahead

    while (fast && fast->next) {
        if (slow == fast) {
            return true; // Cycle detected
        }
        slow = slow->next;
        fast = fast->next->next;
    }
    return false; // No cycle
}

// 2. Hash Table Approach
//   - Uses extra space but is conceptually simple.
//   - Real-world application: Detecting if a file system has a circular directory link (though file systems usually have their own cycle prevention).
bool hasCycle_HashSet(ListNode *head) {
    std::unordered_set<ListNode*> visited;
    ListNode *current = head;
    while (current) {
        if (visited.find(current) != visited.end()) {
            return true; // Cycle detected
        }
        visited.insert(current);
        current = current->next;
    }
    return false; // No cycle
}

// 3. Visited Node Marker (Modifies the list)
//   - Modifies the linked list by changing the 'val' of visited nodes.
//   - Real-world application: Useful in some memory-constrained embedded systems where you can't use extra data structures, and modifying the list is acceptable.  For example, marking nodes during garbage collection in a very simple embedded environment.
bool hasCycle_VisitedMarker(ListNode *head) {
    ListNode *current = head;
    while (current) {
        if (current->val == INT_MIN) { // Using INT_MIN as a marker
            return true; // Cycle detected
        }
        current->val = INT_MIN; // Mark the node as visited
        current = current->next;
    }
    return false; // No cycle
}

// 4. Using a Vector to Store Visited Nodes
//   - Similar to the hash table approach, but uses a vector.  Less efficient for large lists, but shows an alternative.
//   - Real-world application:  Could be adapted for a very small, fixed-size linked list where you want to avoid hash table overhead.  For example, checking for a loop in a short chain of configuration settings in a device.
bool hasCycle_Vector(ListNode *head) {
    std::vector<ListNode*> visited;
    ListNode *current = head;
    while (current) {
        if (std::find(visited.begin(), visited.end(), current) != visited.end()) {
            return true; // Cycle detected
        }
        visited.push_back(current);
        current = current->next;
    }
    return false; // No cycle
}

// 5.  Limited Steps with Check
//    - This approach limits the number of steps taken to traverse the list.
//    - It's not a *guaranteed* cycle detection, but it's useful in situations where you want to prevent infinite loops in a process, even if you can't definitively say a cycle exists.
//    - Real-world application:  A "watchdog timer" in a system that processes linked data.  If the processing takes too long (exceeds the maxSteps), you assume a loop and abort.  This prevents a program from hanging indefinitely.
bool hasCycle_LimitedSteps(ListNode *head, int maxSteps) {
    ListNode *current = head;
    int steps = 0;
    while (current && steps <= maxSteps) {
        current = current->next;
        steps++;
    }
    return (steps > maxSteps); // Return true if it ran for maxSteps.  Indicates *possible* cycle.
}

int main() {
    // Create a linked list with a cycle for testing.
    ListNode *head = new ListNode(1);
    ListNode *node2 = new ListNode(2);
    ListNode *node3 = new ListNode(3);
    ListNode *node4 = new ListNode(4);
    ListNode *node5 = new ListNode(5);

    head->next = node2;
    node2->next = node3;
    node3->next = node4;
    node4->next = node5;
    node5->next = node2; // Create the cycle: 5 points back to 2

    // Test the different cycle detection algorithms.
    std::cout << "Floyd's: " << (hasCycle_Floyd(head) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Hash Set: " << (hasCycle_HashSet(head) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Visited Marker: " << (hasCycle_VisitedMarker(head) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Vector: " << (hasCycle_Vector(head) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Limited Steps (10): " << (hasCycle_LimitedSteps(head, 10) ? "Possible Cycle" : "No Cycle") << std::endl;
    std::cout << "Limited Steps (3): " << (hasCycle_LimitedSteps(head, 3) ? "Possible Cycle" : "No Cycle") << std::endl;

    // Create a linear linked list (no cycle)
    ListNode *linearHead = new ListNode(10);
    ListNode *linearNode2 = new ListNode(20);
    ListNode *linearNode3 = new ListNode(30);
    linearHead->next = linearNode2;
    linearNode2->next = linearNode3;

     std::cout << "Floyd's (Linear): " << (hasCycle_Floyd(linearHead) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Hash Set (Linear): " << (hasCycle_HashSet(linearHead) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Visited Marker (Linear): " << (hasCycle_VisitedMarker(linearHead) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Vector (Linear): " << (hasCycle_Vector(linearHead) ? "Cycle Detected" : "No Cycle") << std::endl;
    std::cout << "Limited Steps (Linear, 10): " << (hasCycle_LimitedSteps(linearHead, 10) ? "Possible Cycle" : "No Cycle") << std::endl;


    // Remember to deallocate the memory, especially if you're running this
    // in an environment where memory management is critical (like an embedded system).
    // In a simple test like this, it's less crucial, but good practice.
    //  delete head;  // Not safe, because the list has a cycle.  Deleting 'head'
    //  would lead to a loop.  You'd need to break the cycle first, or
    //  delete nodes individually in a loop-aware way.
    //  For the linear list:
    delete linearHead;
    delete linearNode2;
    delete linearNode3;

    return 0;
}
